cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/Job.java

```diff
@@ -40,27 +43,27 @@
     /**
      * The Job is awaiting fulfillment of its Trigger.
      */
-    PENDING_TRIGGER,
+    PENDING_LAUNCH(Collections.<State>emptySet()),
     /**
      * The Job is awaiting satisfaction of all of its constraints.
      */
-    PENDING_CONSTRAINT,
+    PENDING_CONSTRAINT(EnumSet.of(PENDING_LAUNCH)),
     /**
      * The Job is awaiting launch.
      */
-    PENDING_LAUNCH;
+    PENDING_TRIGGER(EnumSet.of(PENDING_CONSTRAINT));
 
-    private static final Map<State, State> possibleTransitions = new HashMap<>();
-    static {
-      possibleTransitions.put(PENDING_TRIGGER, PENDING_CONSTRAINT);
-      possibleTransitions.put(PENDING_CONSTRAINT, PENDING_LAUNCH);
+    private final Set<State> allowedNextState;
+
+    State(Set<State> allowedNextState) {
+      this.allowedNextState = allowedNextState;
     }
```

Comment:
State transitions should be defined by what possible states that a state can go to, right? It's part of the state transition definition and is the knowledge of each individual state.

The driver (e.g. constraints checker in this case), should obey the valid state transition rules as defined by each individual state and transit accordingly.

E.g. if we want to define `PENDING_TRIGGER` can go to either `PENDING_CONSTRAINT` or `PENDING_LAUNCH`, then the `allowedNextStates` of `PENDING_TRIGGER` should be `(PENDING_CONSTRAINT, PENDING_LAUNCH)`.